pkg/typesystem/typesystem.go

```diff
@@ -126,39 +126,37 @@ func Difference(base *openfgapb.Userset, sub *openfgapb.Userset) *openfgapb.User
 }
 
 type TypeSystem struct {
-	model           *openfgapb.AuthorizationModel
+	modelID         string
 	schemaVersion   string
 	typeDefinitions map[string]*openfgapb.TypeDefinition
-}
 
-type ValidatedTypeSystem struct {
```

Comment:
The reason for this was so that you can have the Go type system help you pass around a struct that you know has already been validated. Without it it's not clear if a `TypeSystem` struct has already been validated or not.

For example,

```
func blah(ts *ValidatedTypeSystem) { // ts has already been validated }
```
vs
```
func blah(ts *TypeSystem) { // has ts been validated?? }
```

Reply:
this struct is never used though

Reply:
> any time we've called typesystem.New(...).Validate()

this is the only usage

https://github.com/openfga/openfga/blob/89554e8b6a0fb5eca0101bdd9fed3c7d944ebed9/pkg/server/commands/write_authzmodel.go#L49

and it's ignoring the result 😆

Reply:
Well that's not good, lol. If you don't pass model validation then `WriteAuthorizationModel` RPC should fail.

Reply:
Nono, it's ignoring the `validatedTypeSystem` struct, but it's using the `err`

Reply:
Oh, I see. Yeah, so ideally we propagate the `ValidatedTypeSystem` anywhere we assumed we are working with a pre-validated typesystem. That was the goal, but it seems when we added that we didn't properly wire it up??

Reply:
you mean that all the other APIs would go from accepting a `typesystem` to accepting a `ValidTypeSystem`? but what does that mean, it was valid at the time it was written for sure, but what if the validation rules changed after that?

Reply:
What if we had the constructor do the validation so that we know any `*TypeSystem` returned has already been validated. Basically, I want to just ensure that

```go
typesys, err := typesystem.New(model) // should always return a validated typesystem
```
and we could have a helper if needs be like we do elsewhere
```go
typesys := typesystem.MustNew(model) // panics if the typesystem failed validation during construction.
```
then we could just get rid of the `ValidatedTypeSystem` in favor of that.

Reply:
We can send a future PR with that, but I need to think about your idea a bit more. Off the top of my head

> What if we had the constructor do the validation

- Check for example should not do validations, to keep it fast.
- In general i think `New`s should be kept as thin as possible as not throw any errors, otherwise it becomes `NewAndValidate`.